### PR TITLE
scaffolder/next: start of the new `OngoingTask` page

### PR DESCRIPTION
**Pull request by benjdlambert**
This is a quick pass of the `OngoingTask` page so that we can start to test the `scaffolder/next` work end to end in upcoming releases.

Screenshot 2023-01-31 at 14 08 12

image

### CODE REVIEW

File: packages/core-components/src/components/LogViewer/LogLine.tsx [hunk 1]
```diff
@@ -159,8 +158,8 @@ export function LogLine({
   const elements = useMemo(
     () =>
       chunks.map(({ text, modifiers, highlight }, index) => (
-        <Typography
-          component="span"
+        // eslint-disable-next-line react/forbid-elements
```
Comment (benjdlambert) [inline packages/core-components/src/components/LogViewer/LogLine.tsx:161]: This is a slightly unrelated fix, that the styles are really all wrong when you wrap it in a mui element. So went back to how it was before.

**Comment by acierto** [comment]
Looks nice!

@benjdlambert Will a new version allow to see the summary of already executed task?
Like the last step what you see prior to start the task?

Currently i'm missing that info, when i'm opening already executed task, i have to search in logs what have i chosen for this run

**Comment by acierto** [comment]
How it will scale with 20 steps for example? 😅

File: plugins/scaffolder/src/next/OngoingTask/TaskSteps/StepTime.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    const startedAt = DateTime.fromISO(step.startedAt);
+    const formatted = Interval.fromDateTimes(startedAt, end)
+      .toDuration()
+      .valueOf();
+
+    setTime(humanizeDuration(formatted, { round: true }));
+  }, [step.endedAt, step.startedAt]);
+
+  useMemo(() => calculate(), [calculate]);
```
Comment (freben) [inline plugins/scaffolder/src/next/OngoingTask/TaskSteps/StepTime.tsx:51]: Matter of taste, but how do you like this

```ts
const [counter, setCounter] = useState(0);

const time = useMemo(
  () => { ...compute... },
  [step.startedAt, step.endedAt, (!step.endedAt && counter)]
);

useInterval(() => setCounter(c => c + 1), 1000);
```

it just feels odd to (1) depend on the instance value of a function, (2) use useMemo for what's effectively an effect, and (3) have a callback to effectively do setState, as a way of achieving memoization instead of using useMemo

File: plugins/scaffolder/src/next/OngoingTask/TaskSteps/StepTime.test.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    const step = {
+      id: 'test',
+      startedAt: undefined,
+      endedAt: undefined,
+      name: 'test',
+    };
+
+    const { container } = render(<StepTime step={step} />);
+
+    expect(container.querySelector('span')?.textContent).toBe('');
```
Comment (freben) [inline plugins/scaffolder/src/next/OngoingTask/TaskSteps/StepTime.test.tsx:31]: just a nudge that `screen` is a great way of doing these queries in general, so you don't ever have to care about the return values from `render`.

File: plugins/scaffolder/src/next/OngoingTask/Outputs/DefaultOutputs.tsx [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+export const DefaultOutputs = (props: { output?: ScaffolderTaskOutput }) => {
+  if (!props.output?.links) {
+    return null;
+  }
+
+  return (
+    <Box paddingBottom={2}>
+      <Paper>
+        <Box padding={2} justifyContent="center" display="flex" gridGap={16}>
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/scaffolder/src/next/OngoingTask/Outputs/DefaultOutputs.tsx:29]: It is indeed.

File: plugins/scaffolder/src/next/OngoingTask/Outputs/LinkOutputs.tsx [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        })
+        .map(({ url, title, icon }, i) => {
+          const Icon = iconResolver(icon);
+          return (
+            <Link
+              classes={{ root: classes.root }}
+              href={url}
+              key={i}
+              component={Button}
+              target="_blank"
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/scaffolder/src/next/OngoingTask/Outputs/LinkOutputs.tsx:61]: core components doesn't accept the `component` prop. Maybe I can look into adding that.

File: plugins/scaffolder/src/next/OngoingTask/TaskSteps/StepTime.tsx (hunk 2 above)
Comment (benjdlambert) [reply plugins/scaffolder/src/next/OngoingTask/TaskSteps/StepTime.tsx:51]: i'm feeling it might be better off to work out if we can drop the `useMemo` entirely for readability sake. Like `useMount(() => calculate())` or something

File: plugins/scaffolder/src/next/OngoingTask/TaskSteps/TaskSteps.test.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import { renderInTestApp } from '@backstage/test-utils';
+
+describe('TaskSteps', () => {
+  it('should render each of the steps', async () => {
+    const steps = [
+      {
+        id: '1',
+        name: 'Step 1',
+        status: 'processing' as ScaffolderTaskStatus,
+        startedAt: Date.now().toLocaleString(),
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/scaffolder/src/next/OngoingTask/TaskSteps/TaskSteps.test.tsx:28]: Not this test, as it doesn't care about the startedAt time, it's just checking the labels and the steps are rendered. The one that tests the actual rendering of the time uses `jest.useMockTimers()`

File: plugins/scaffolder/src/next/OngoingTask/TaskSteps/TaskSteps.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+} from '@material-ui/core';
+import { TaskStep } from '@backstage/plugin-scaffolder-common';
+import { Step } from '../../../components/hooks/useEventStream';
+import { StepIcon } from './StepIcon';
+import { StepTime } from './StepTime';
+
+interface StepperProps {
+  steps: (TaskStep & Step)[];
+  activeStep?: number;
+  setActiveStep?: (step: number) => void;
```
[the remark this answers is not in the document]
Comment (benjdlambert) [reply plugins/scaffolder/src/next/OngoingTask/TaskSteps/TaskSteps.tsx:32]: It's not, this was just copied from the other component for now. It probably will be used as it's basically what @acierto mentions above to filter the logs for that step only.

File: plugins/scaffolder/src/next/routes.ts [hunk 8]
```diff
@@ -34,3 +34,24 @@ export const nextScaffolderTaskRouteRef = createSubRouteRef({
...
+
+/** @alpha */
+export const nextActionsRouteRef = createSubRouteRef({
+  id: 'scaffolder/next/actions',
+  parent: nextRouteRef,
+  path: '/actions',
+});
+
+/** @alpha */
+export const nextEditRouteRef = createSubRouteRef({
```
Comment (benjdlambert) [inline plugins/scaffolder/src/next/routes.ts:53]: This is still TODO, need to move the TemplateEditor to the new version of `react-jsonschema-form`. Followup PR that can wait until next week.

File: plugins/scaffolder/src/next/OngoingTask/ContextMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+  Popover,
+} from '@material-ui/core';
+import Retry from '@material-ui/icons/Repeat';
+import Toc from '@material-ui/icons/Toc';
+import MoreVert from '@material-ui/icons/MoreVert';
+import React, { useState } from 'react';
+
+type ContextMenuProps = {
+  logsVisible?: boolean;
+  onToggleLogs?: (state: boolean) => void;
```
Comment (freben) [inline plugins/scaffolder/src/next/OngoingTask/ContextMenu.tsx:33]: onSetLogsVisible might be a nice pairing with the above

**Review by freben — APPROVED** [review]
